codalab/lib/bundle_store.py

```diff
@@ -94,74 +81,79 @@ def upload(self, sources, follow_symlinks, exclude_patterns, git, unpack, remove
         '''
         to_delete = []
 
-        # Create temporary directory as a staging area and put everything there.
-        temp_path = tempfile.mkdtemp('-bundle_store_upload')
-        temp_subpaths = []
+        # If just a single file, set the final path to be equal to that file
+        single_file = len(sources) == 1
+
+        final_path = os.path.join(self.data, uuid)
+        if os.path.exists(final_path):
+            raise UsageError('Error, path %s already present in bundle store' % final_path)
+        # Only make if not there
+        elif not single_file:
+            path_util.make_directory(final_path)
+
+        # Paths to resources
+        subpaths = []
+
         for source in sources:
             # Where to save |source| to (might change this value if we unpack).
-            temp_subpath = os.path.join(temp_path, os.path.basename(source))
+            if not single_file:
+                subpath = os.path.join(final_path, os.path.basename(source))
+            else:
+                subpath = final_path
+
             if remove_sources:
                 to_delete.append(source)
```

Comment:
I don't think we need `to_delete` anymore.  This was when we were first uploading to a temporary directory and then moving it to the bundle store (see remote_bundle_client's upload method).  But now, uploading should just go directly to the bundle store.  This might require testing the upload / download / add bundle functionality carefully.